### PR TITLE
OCLOMRS-532: When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary

### PR DESCRIPTION
**Pull request by myrdstom**
# JIRA TICKET NAME:
[When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary](https://issues.openmrs.org/browse/OCLOMRS-532)

# Summary:
When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary

### CODE REVIEW

**Comment by HerbertYiga** [comment]
Have you looked through the failing CI logs

**Comment by myrdstom** [comment]
> Have you looked through the failing CI logs

The PR is a work in progress. I am aware of the failing CI logs. When it is ready for review, the CI logs won't be failing

**Comment by myrdstom** [comment]
> Have you looked through the failing CI logs

Yes, I have

File: src/components/bulkConcepts/component/ActionButtons.jsx [hunk 1]
```diff
@@ -40,14 +41,21 @@ export class ActionButtons extends Component {
   addConcept = (conceptUrl, name) => {
     const { params } = this.props;
     const data = { data: { expressions: [conceptUrl] } };
-    this.props.addConcept(params, data, name);
+    if (this.props.mappings && this.props.mappings.length > 0) {
+      this.props.addConcept(params, data, name);
+      this.props.mappings.map((concept) => {
```
Comment (Karuhanga) [inline src/components/bulkConcepts/component/ActionButtons.jsx]: Is there a way of adding multiple references to concepts in one go? This seems like it will not scale well if we keep making an api call for each mapping.

Comment (myrdstom) [reply src/components/bulkConcepts/component/ActionButtons.jsx]: You are right Lincoln, multiple API calls could slow down the application. I will figure out a work around

**Comment by dkayiwa** [comment]
Are all the changes in bulkConcept.test.js related to this ticket?

**Comment by myrdstom** [comment]
> Are all the changes in bulkConcept.test.js related to this ticket?
Yes, what I am going to do is refactor the test to make it more readable

**Comment by dkayiwa** [comment]
Shouldn't we create a separate ticket for refactoring? Your commit has too many changes which is not good.

**Comment by myrdstom** [comment]
> Shouldn't we create a separate ticket for refactoring? Your commit has too many changes which is not good.

Ok, will do. I will get back to you when the ticket is ready for review

**Comment by myrdstom** [comment]
> Shouldn't we create a separate ticket for refactoring? Your commit has too many changes which is not good.

The Refactoring PR has been merged as suggested

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 2]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
     wrapper.find('#add-button').simulate('click');
     jest.runAllTimers();
   });
-
+  it('should check whether recursiveQuestionAnswers has a length greater than one', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Are you sure it is greater than one?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Sorry about that, I have rectified this

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 3]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
     wrapper.find('#add-button').simulate('click');
     jest.runAllTimers();
   });
-
+  it('should check whether recursiveQuestionAnswers has a length greater than zero', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Is this test really about ensuring that the tested code checks for length to be greater than zero?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: I have adjusted the description accordingly

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 4]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
     wrapper.find('#add-button').simulate('click');
     jest.runAllTimers();
   });
-
+  it('should check whether the concept referenced in recursiveQuestionAnswers has a mapping', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Is it the test that should check whether recursiveQuestionAnswers has a mapping? Or is the test ensuring that the tested code does so?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: The test ensures that the tested code has a mapping. let me rephrase the description accordingly

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Done, is the new description satisfactory?

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Can you share this with your team mates?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: I have had a discussion with the teammates and adjusted the description accordingly, does it suffice?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: I have had a discussion with the teammates and adjusted the description accordingly, does it suffice?

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 5]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
...
+      </Router>
+    </Provider>);
+    wrapper.find('#add-button').simulate('click');
+    expect(
+      wrapper.instance().props.children.props.children.props.recursiveQuestionAnswers[0]
+        .mappings.length,
+    ).toEqual(1);
+    jest.runAllTimers();
+  });
+  it('should handle CIEL concepts without mappings', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Are the answers the same as mappings?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Yes, answers are a subset of mappings. Another example of a mappings subset is `CONCEPT-SET`. A clear example of this can be seen in this [link](https://qa.openconceptlab.org/orgs/CIEL/sources/CIEL/concepts/1268/) and this [link](https://qa.openconceptlab.org/orgs/CIEL/sources/CIEL/concepts/1085/).
This particular PR handles the addition of all various types of mappings for a concept.

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 6]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
     wrapper.find('#add-button').simulate('click');
     jest.runAllTimers();
   });
-
+  it('should handle CIEL concepts with mappings', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: What do you mean by "should handle"?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: I have adjusted the description. It specifies that the test checks the code to confirm whether a concept has any mappings.

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js [hunk 7]
```diff
@@ -297,7 +297,58 @@ describe('Test suite for BulkConceptsPage component', () => {
     wrapper.find('#add-button').simulate('click');
     jest.runAllTimers();
   });
-
+  it('should check whether the CIEL concept has answer mappings', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: You have multiple instances of should handle, down below.
Secondly, can you point me to the code, being tested (not the test), that checks whether CIEL concepts have mappings?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: > You have multiple instances of should handle, down below.

I believe you looked at the code as I was making changes. IT should have been rectified
 
> Secondly, can you point me to the code, being tested (not the test), that checks whether CIEL concepts have mappings?

`src/components/bulkConcepts/component/ActionButtons.jsx`. Lines 47 to 67

File: src/components/bulkConcepts/component/ActionButtons.jsx [hunk 8]
```diff
@@ -40,14 +44,42 @@ export class ActionButtons extends Component {
   addConcept = (conceptUrl, name) => {
     const { params } = this.props;
     const data = { data: { expressions: [conceptUrl] } };
-    this.props.addConcept(params, data, name);
```
Comment (myrdstom) [inline src/components/bulkConcepts/component/ActionButtons.jsx]: Lines 47 to 67

File: src/components/bulkConcepts/component/ActionButtons.jsx [hunk 9]
```diff
@@ -40,14 +44,42 @@ export class ActionButtons extends Component {
...
+    if (this.props.mappings && this.props.mappings.length > 0) {
+      this.props.addConcept(params, data, name);
```
Comment (myrdstom) [inline src/components/bulkConcepts/component/ActionButtons.jsx]: Lines 47 to 67

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js (hunk 7 above)
Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: > You have multiple instances of should handle, down below.

I believe you looked at the code as I was making changes. IT should have been rectified
 
> Secondly, can you point me to the code, being tested (not the test), that checks whether CIEL concepts have mappings?

`src/components/bulkConcepts/component/ActionButtons.jsx`.

File: src/tests/bulkConcepts/reducer/index.test.js [hunk 10]
```diff
@@ -149,4 +153,30 @@ describe('Test suite for bulkConcepts reducer', () => {
       preview: action.payload,
     });
   });
+  it('should handle GET_QUESTION_ANSWERS', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/reducer/index.test.js]: What does "should handle" mean?

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js (hunk 7 above)
Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Is this test about ensuring that when a concept does not have answer mappings then it should not save it?

File: src/tests/bulkConcepts/reducer/index.test.js (hunk 10 above)
Comment (myrdstom) [reply src/tests/bulkConcepts/reducer/index.test.js]: I have omitted the multiple instances of `Should handle`

File: src/tests/bulkConcepts/containers/BulkConceptsPage.test.js (hunk 7 above)
Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: No, this test is about checking whether a concept has an answer mapping or not. The logic that handles saving of the information is in a completely separate file

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: What happens when it does not have an answer mapping?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Additionally, if a concept has a mapping, it should be saved. If it does not have a mapping, it should still be saved

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Then in that case, what is the use of this test?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: This test simply confirms that the code we have written to check that whether an incoming concept has a mapping or not is working as intended.

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Can you discuss this with your team mates?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Sorry, I edited the comment and had not seen the above message. Could you confirm whether it my explanation is satisfactory?

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: You test description does not reflect that

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: I liaised with the team and have edited the description to be more specific to checking how the recursive methods handle the concept data they receive. Given that those were the missing lines and this is exactly what the test is testing, is the description sufficient?

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Before i take a second look at the description, does the test fail, if you remove the lines (Lines 47 to 67 in ActionButtons) that you claim to be testing?

Comment (myrdstom) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: No, let me look into this and get back to you

Comment (dkayiwa) [reply src/tests/bulkConcepts/containers/BulkConceptsPage.test.js]: Then it is not testing anything.
